fiaas_deploy_daemon/deployer/kubernetes/deployment/deployer.py

```diff
@@ -121,7 +121,7 @@ def deploy(self, app_spec, selector, labels, besteffort_qos_is_required):
                               strategy=deployment_strategy)
 
         deployment = Deployment.get_or_create(metadata=metadata, spec=spec)
-        self._datadog.apply(deployment, app_spec, besteffort_qos_is_required)
+        self._datadog.apply(deployment, app_spec, besteffort_qos_is_required, self._grace_period)
```

Comment:
Is it possible to use `self._lifecycle` from https://github.com/fiaas/fiaas-deploy-daemon/blob/master/fiaas_deploy_daemon/deployer/kubernetes/deployment/deployer.py#L51 instead of `self._grace_period` so we ensure the same pre-stop is being used in the main and sidecar containers?

This way the datadog container will also comply with the 30 seconds minimum grace period for a graceful shutdown of the container if necessary after the SIGTERM (and before SIGKILL).

[This](https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-terminating-with-grace) is a nice short writeup on how this process works.

Reply:
The point of using `self._grace_period` here is to ensure that the datadog side car stays alive for as long as the main application container itself is.

Using `self._lifecycle`, as you suggest, means that both the datadog sidecar and the main application container would receive `SIGTERM` at the same time. From my limited experience, the datadog container exits within 1s of receiving `SIGTERM`, while the main application container could be running for another 30s before receiving a final `SIGKILL` and exiting. 

So, the question really is this: how much effort do we want to put into ensuring metrics emitted by the main container are preserved?

If we don't care about preserving metrics from the moment `SIGTERM` is issued, then `self._lifecycle` could be shared between the containers.

If, on the other hand, we want to do everything we can to preserve emitted metrics, then I think we want to take the full grace period into account, and keep it as I have it here.

Reply:
When using the grace period like this, you force all pods using datadog to wait out their grace period before shutting down. This will make deployment way slower than it needs to be. The pod doesn't stop before all containers have exited.

Reply:
Instead of giving the datadog sidecar the full grace period, how would you feel about giving the datadog sidecar a few more seconds? Say, we'd wait `pre_stop_delay + 5` seconds before sending the sidecar a `SIGTERM`?

That would allow well behaved applications that exit quickly an opportunity to send metrics on termination. It would offer no guarantees to applications that otherwise take "too long" to exit.

Reply:
+1 to Joao's last comment. I think it's a reasonable trade-off to wait 5s after pre-stop.  K8s 1.18 introduces sidecar lifecycle hooks and ordering, so we can iterate/improve on this problem later.

Reply:
I've updated my changes so that the datadog sidecar gets a pre-stop delay based on the main container's value with the proposed additional 5 seconds.

Let me know what you think.